pkg/logging/logger.go

```diff
@@ -0,0 +1,83 @@
+package logging
+
+import (
+	"fmt"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apiserver/pkg/storage/names"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+// Logger
+// Delegates functionality to the wrapped `Real` logger.
+type Logger struct {
+	Real    logr.Logger
+	history map[error]bool
+	name    string
+}
+
+// Get a named logger.
+func WithName(name string) Logger {
+	logger := Logger{
+		Real: logf.Log.WithName(name),
+		name: name,
+	}
+	logger.Reset()
+	return logger
+}
+
+// Reset the logger.
+// Updates the generated correlation suffix in the name and
+// clears the reported error history.
+func (l *Logger) Reset() {
+	name := fmt.Sprintf("%s|", l.name)
+	name = names.SimpleNameGenerator.GenerateName(name)
+	l.Real = logf.Log.WithName(name)
+	l.history = make(map[error]bool)
+}
+
+// Logs at info.
+func (l Logger) Info(message string, kvpair ...interface{}) {
+	l.Real.Info(message, kvpair...)
+}
+
+// Logs an error.
+// Previously logged errors are ignored.
+func (l Logger) Error(err error, message string, kvpair ...interface{}) {
+	if err == nil {
+		return
+	}
+	_, found := l.history[err]
+	if found || errors.IsConflict(err) || errors.IsNotFound(err) {
```

Comment:
Just to point this out to anyone looking at this PR: this new logger will skip printing the error for any log.Error() calls for:

- Previously printed identical (same pointer) error messages from the same reconcile

- `Conflict` (resourceVersion not as expected during update attempt, resolved by reconcile retry)

- `NotFound` (resource which was requested not found, which is expected in many places. @jortel I guess this would also prevent us from printing error messages when something is legitimately not found though? E.g. a situation where something we were expecting to exist got deleted somewhere along the way. What did you envision for dealing with a situation like I'm describing?

Reply:
> @jortel I guess this would also prevent us from printing error messages when something is legitimately not found though? E.g. a situation where something we were expecting to exist got deleted somewhere along the way. What did you envision for dealing with a situation like I'm describing?

Sort of.  All of our _ensure_ logic does a  `Get()` on resources that may/may-not already exists.  The not-found is an anticipated, non-error result.